tests/std/tests/VSO_0000000_allocator_propagation/test.cpp

```diff
@@ -34,6 +34,165 @@ _CONSTEXPR20 void assert_is_permutation(const Container& cont, initializer_list<
     assert(is_permutation(cont.begin(), cont.end(), il.begin(), il.end()));
 }
 
+template <class T>
+struct lifetime_checked_pointer {
+    T* ptr;
+    bool living;
+
+    _CONSTEXPR20 lifetime_checked_pointer() noexcept : ptr(nullptr), living(true) {}
+
+    _CONSTEXPR20 explicit lifetime_checked_pointer(T* ptr_) noexcept : ptr(ptr_), living(true) {}
+    _CONSTEXPR20 /*implicit*/ lifetime_checked_pointer(nullptr_t) noexcept : ptr(nullptr), living(true) {}
+
+    _CONSTEXPR20 lifetime_checked_pointer(const lifetime_checked_pointer& other) noexcept
+        : ptr(other.ptr), living(other.living) {
+        assert(living);
+    }
+    template <class U, enable_if_t<is_convertible_v<U*, T*>, int> = 0>
+    _CONSTEXPR20 lifetime_checked_pointer(const lifetime_checked_pointer<U>& other) noexcept
+        : ptr(other.ptr), living(other.living) {
+        assert(living);
+    }
+
+    _CONSTEXPR20 ~lifetime_checked_pointer() {
+        assert(living);
+        living = false;
+    }
+
+    _CONSTEXPR20 lifetime_checked_pointer& operator=(const lifetime_checked_pointer& other) noexcept {
+        assert(living && other.living);
+        ptr = other.ptr;
+        return *this;
+    }
+
+    _CONSTEXPR20 explicit operator bool() const noexcept {
+        assert(living);
+        return static_cast<bool>(ptr);

```

Comment:
strangely, when I uses `ptr != nullptr`, I get the following error:

<details>
<summary>Result</summary>

```
-- Testing: 30 tests, 16 workers --
PASS: std :: tests/VSO_0000000_allocator_propagation:00 (1 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:08 (2 of 30)
UNSUPPORTED: std :: tests/VSO_0000000_allocator_propagation:17 (3 of 30)
UNSUPPORTED: std :: tests/VSO_0000000_allocator_propagation:18 (4 of 30)
UNSUPPORTED: std :: tests/VSO_0000000_allocator_propagation:19 (5 of 30)
UNSUPPORTED: std :: tests/VSO_0000000_allocator_propagation:20 (6 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:10 (7 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:09 (8 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:12 (9 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:01 (10 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:14 (11 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:21 (12 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:07 (13 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:22 (14 of 30)
UNSUPPORTED: std :: tests/VSO_0000000_allocator_propagation:29 (15 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:13 (16 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:15 (17 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:05 (18 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:04 (19 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:23 (20 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:06 (21 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:02 (22 of 30)
FAIL: std :: tests/VSO_0000000_allocator_propagation:24 (23 of 30)
******************** TEST 'std :: tests/VSO_0000000_allocator_propagation:24' FAILED ********************
Build setup steps:
Build steps:
Build step failed unexpectedly.
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.33030\bin\HostX64\x64\cl.EXE" "-c" "(replaced)\STL\tests\std\tests\VSO_0000000_allocator_propagation\test.cpp" "-I(replaced)\STL\out\x64\out\inc" "-I(replaced)\STL\llvm-project\libcxx\test\support" "-I(replaced)\STL\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/w15267" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/BE" "/c" "/EHsc" "/MTd" "/std:c++latest" "/permissive-" "/w14640" "/Zc:threadSafeInit-"
Exit Code: 2
Standard Output:
--
test.cpp
"(replaced)\STL\tests\std\tests\VSO_0000000_allocato
          r_propagation\test.cpp", line 1844: error: expression must have a
          constant value
      static_assert(test_sequence<vector>());
                    ^
"(replaced)\STL\tests\std\tests\VSO_0000000_allocato
          r_propagation\test.cpp", line 70: note: cannot compare a pointer past
          the end of an array with a pointer to another complete object
          return ptr != nullptr;
                     ^
"(replaced)\STL\out\x64\out\inc\vector", line 111:
          note: called from:
          _STL_VERIFY(_Off == 0 || _Ptr, "cannot seek value-initialized vector iterator");
          ^
"(replaced)\STL\out\x64\out\inc\vector", line 124:
          note: called from:
          _Verify_offset(_Off);
                        ^
"(replaced)\STL\out\x64\out\inc\vector", line 305:
          note: called from:
          _Mybase::operator+=(_Off);
                             ^
"(replaced)\STL\out\x64\out\inc\xutility", line
          1414: note: called from:
          _Where += _Off;
                 ^
"(replaced)\STL\out\x64\out\inc\xutility", line
          1478: note: called from:
      _STD advance(_First, -_Off);
                  ^
"(replaced)\STL\tests\std\tests\VSO_0000000_allocato
          r_propagation\test.cpp", line 272: note: called from:
      *prev(src.end()) += 8;
           ^
"(replaced)\STL\tests\std\tests\VSO_0000000_allocato
          r_propagation\test.cpp", line 459: note: called from:
      test_sequence_copy_ctor<Sequence>();
                                       ^

--


********************
PASS: std :: tests/VSO_0000000_allocator_propagation:03 (24 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:26 (25 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:25 (26 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:11 (27 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:27 (28 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:28 (29 of 30)
PASS: std :: tests/VSO_0000000_allocator_propagation:16 (30 of 30)
********************
Failed Tests (1):
  std :: tests/VSO_0000000_allocator_propagation:24


Testing Time: 21.60s
  Unsupported:  5
  Passed     : 24
  Failed     :  1
```
</details>

Reply:
Perhaps we should add some comments for EDG...

Reply:
Do you mind minimizing this so we can report the bug to EDG?

Reply:
Yeah, here it is:
`"/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/w15267" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/BE" "/c" "/EHsc" "/MTd" "/std:c++latest" "/permissive-" "/w14640" "/Zc:threadSafeInit-"`
```cpp
#include<cassert>

template<class T>
struct ptr {
    T* p;

    constexpr explicit operator bool() {
        //return static_cast<bool>(p); // ok.
        return p != nullptr; // problematic.
    }
};


constexpr bool problematic() {
    int arr[10]{};
    ptr p{ arr + 10 };
    assert(p);

    return true;
}

int main() {
    static_assert(problematic());
}
```

Reply:
Awesome - reported as VSO-1888157 to the MSFT-internal EDG team.

Reply:
```suggestion
#ifdef __EDG__ // TRANSITION, VSO-1888157
        return static_cast<bool>(ptr);
#else // ^^^ workaround / no workaround vvv
        return ptr != nullptr;
#endif // ^^^ no workaround ^^^
```

Reply:
> Awesome - reported as [VSO-1888157](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1888157) to the MSFT-internal EDG team.

Late notification: this was fixed for VS 2022 17.9.